tensorflow/cc/tools/freeze_saved_model.cc

```diff
@@ -204,16 +227,35 @@ Status FreezeGraphDef(const SavedModelBundle& saved_model_bundle,
     if (variable_node_names.find(node.name()) != variable_node_names.end()) {
       ConvertVariableToConstant(node, variable_to_value_map[node.name()],
                                 frozen_graph_def->add_node());
-    } else if (node.op() == "ReadVariableOp" &&
-               variable_node_names.find(node.input(0)) !=
-                   variable_node_names.end()) {
-      // If the node is a ReadVariableOp, its input VarHandleOp will be
-      // converted to a Constant, so we will need to convert it to an Identity.
-      ConvertReadVariableOpToIdentity(node, frozen_graph_def->add_node());
-    } else {
-      // If the node isn't a variable, just copy the node as-is.
-      *frozen_graph_def->add_node() = node;
+      continue;
+    } else if (node.op() == "ReadVariableOp" || node.op() == "Identity") {
+      StatusOr<string> status_or_name =
+          GetVarHandleName(name_to_node_map, node.name());
+      if (status_or_name.ok()) {
+        string var_handle_name = status_or_name.ValueOrDie();
+        if (variable_node_names.find(var_handle_name) !=
+            variable_node_names.end()) {
+          if (node.op() == "ReadVariableOp") {
+            // If the node is a ReadVariableOp, its input VarHandleOp will be
+            // converted to a Constant, so we will need to convert it to an
+            // Identity.
+            ConvertReadVariableOpToIdentity(node, frozen_graph_def->add_node());
+            continue;
+          } else if (node.op() == "Identity") {
+            // Identity node that is forwarding the value of a frozen
+            // VarhandleOp. We ensure that the dtype matches of the variable
+            // dtype.
+            NodeDef* new_node = frozen_graph_def->add_node();
+            new_node->CopyFrom(node);
+            (*new_node->mutable_attr())["T"] =
+                name_to_node_map.at(var_handle_name)->attr().at("dtype");
```

Comment:
I don't quite understand this code. It create new_node and set up the "T" attribute of the new_node.
But the new_node is not used to replace "node" and the "T" attribute of "node" is not changed.
So how we will use the information in "new_node"?